### PR TITLE
poetry for dependencies fixes #3393

### LINKED ISSUE #3393: Use Poetry for python dependency management
[Poetry](https://python-poetry.org/) is the new coolness for managing python dependencies and automatically manages virtual environments so I think switching over to this could make managing local dev environments easier, not to mention doing away with the annoying hash workflow, so let's switch over to this.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-253)

### PR DESCRIPTION
**Pull request by tiftran**

### CODE REVIEW

File: app/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+pyparsing = "2.4.7"
+django-markdown2 = "0.3.1"
+markdown2 = "2.3.9"
+tblib = "1.7.0"
+pytest-xdist = "1.34.0"
+djangorestframework-csv = "2.1.0"
+unicodecsv = "0.14.1"
+kinto-http = "10.7.0"
+jsonschema = "3.2.0"
+mozilla-nimbus-shared = "0.0.9"
```
[the remark this answers is not in the document]
Comment (jaredlockhart) [reply app/pyproject.toml:52]: Dependabot says it will: https://dependabot.com/blog/announcing-poetry-support/

**Comment by jaredlockhart** [comment]
Awesome @tiftran !

I got it working on my mac using the following commands:

```
brew install postgresql llvm openssl yarn

echo 'export PATH="/usr/local/opt/llvm/bin:$PATH"' >> ~/.bash_profile
export LIBRARY_PATH=$LIBRARY_PATH:/usr/local/opt/openssl/lib/

poetry install

osx catalina, reinstall command line tools:
https://medium.com/flawless-app-stories/gyp-no-xcode-or-clt-version-detected-macos-catalina-anansewaa-38b536389e8d

yarn install
poetry shell

yarn workspace @experimenter/core build

.env:
DEBUG=True
POSTGRES_HOST=localhost
HOSTNAME=localhost

./manage.py runserver 0.0.0.0:7001
```

We should update the README.md with the setup instructions on how to get this working.  Add a section next to the docker setup instructions with the complete instructions to get a poetry dev environment working for mac.  We can add ones for windows/linux when people try it and document their steps.

File: app/Dockerfile [hunk 2]
```diff
@@ -18,13 +18,23 @@ RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources
...
 # Python packages
-COPY ./requirements /app/requirements
-RUN pip install -r requirements/default.txt --no-cache-dir --disable-pip-version-check
+RUN curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python
+ENV PATH "/root/.poetry/bin:${PATH}"
+
+# Make and activate a Python virtualenv
+RUN python -m venv /opt/venv
+ENV PATH "/opt/venv/bin:${PATH}"
+ENV VIRTUAL_ENV="/opt/venv"
```
Comment (jaredlockhart) [inline app/Dockerfile:30]: Oh neat, this worked, but I also discovered you can just disable poetry venv mode altogether with:

```
RUN poetry config virtualenvs.create false

# # Make and activate a Python virtualenv
# RUN python -m venv /opt/venv
# ENV PATH "/opt/venv/bin:${PATH}"
# ENV VIRTUAL_ENV="/opt/venv"
```

and that worked for me too, maybe that's simpler?

File: app/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+[tool.poetry.dev-dependencies]
```
Comment (jaredlockhart) [inline app/pyproject.toml:54]: Oh sweet, this'll let us move dev dependencies in here.  Maybe someday we can do that thing where we create two docker images, one for development and one for production.  Prolly can do that outside this pr though.

File: app/Dockerfile [hunk 4]
```diff
@@ -18,13 +18,23 @@ RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources
...
+
+COPY poetry.lock pyproject.toml ./
+RUN poetry install
+
 
 # If any package is installed, that is incompatible by version, this command
-# will exit non-zero and print what is usually just a warning in `pip install`
```
[the remark this answers is not in the document]
Comment (tiftran) [reply app/Dockerfile:26]: oops, will revert that

File: app/pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tiftran) [reply app/pyproject.toml:52]: I've tested out using my fork and it works! ex https://github.com/tiftran/experimenter/pull/24/files

**Review by pdehaan — APPROVED** [review]
Superficially looks good to me (but a bit out of my depth) -- but excited to give it all a try on stage.
Looks like it needs a rebasin'.

Will leave it for @jaredlockhart to give final 👀 and 👍 .

**Comment by jaredlockhart** [comment]
@tiftran Still needs docs in the README?

**Comment by tiftran** [comment]
@jaredlockhart  right, right but was thinking maybe https://github.com/mozilla/experimenter/pull/3420 should be merged first before the readme is updated? since the readme update, you mentioned includes running experimenter locally and poetry?

**Comment by jaredlockhart** [comment]
OOPS i merged this by mistake XD

okay I'll merge https://github.com/mozilla/experimenter/pull/3420

and we can update the readme as a followup: https://github.com/mozilla/experimenter/issues/3436

**Pull request merged**